src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VCProjectSystemServices.cs

```diff
@@ -0,0 +1,407 @@
+                GeneratePathProperty = MSBuildStringUtility.IsTrue(GetReferenceMetadataValue(reference, ProjectItemProperties.GeneratePathProperty)),
+                LibraryRange = new LibraryRange(
+                    name: reference.Name,
+                    versionRange: VersionRange.Parse(reference.Version),
+                    typeConstraint: LibraryDependencyTarget.Package)
+            };
+
+            MSBuildRestoreUtility.ApplyIncludeFlags(
+                dependency,
+                GetReferenceMetadataValue(reference, ProjectItemProperties.IncludeAssets),
+                GetReferenceMetadataValue(reference, ProjectItemProperties.ExcludeAssets),
+                GetReferenceMetadataValue(reference, ProjectItemProperties.PrivateAssets));
+
+
+            // Add warning suppressions
+            foreach (var code in MSBuildStringUtility.GetNuGetLogCodes(GetReferenceMetadataValue(reference, ProjectItemProperties.NoWarn)))
+            {
+                dependency.NoWarn.Add(code);
+            }
+
+            return dependency;
+        }
+
+        private static string GetReferenceMetadataValue(PackageReference reference, string metadataElement)
+        {
+            Assumes.Present(reference);
+            Assumes.NotNullOrEmpty(metadataElement);
+
+            if (reference.MetadataElements == null || reference.MetadataValues == null)
+            {
+                return string.Empty; // no metadata for package
+            }
+
+            var index = Array.IndexOf(reference.MetadataElements, metadataElement);
+            if (index >= 0)
+            {
+                return reference.MetadataValues.GetValue(index) as string;
+            }
+
+            return string.Empty;
+        }
+
+        public async Task AddOrUpdatePackageReferenceAsync(LibraryDependency packageReference, CancellationToken _)
+        {
+            Assumes.Present(packageReference);
+
+            await _threadingService.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+            var includeFlags = packageReference.IncludeType;
+            var privateAssetsFlag = packageReference.SuppressParent;
+            var metadataElements = new List<string>();
+            var metadataValues = new List<string>();
+            if (includeFlags != LibraryIncludeFlags.All)
+            {
+                metadataElements.Add(ProjectItemProperties.IncludeAssets);
+                metadataValues.Add(LibraryIncludeFlagUtils.GetFlagString(includeFlags).Replace(',', ';'));
+            }
+
+            if (privateAssetsFlag != LibraryIncludeFlagUtils.DefaultSuppressParent)
+            {
+                metadataElements.Add(ProjectItemProperties.PrivateAssets);
+                metadataValues.Add(LibraryIncludeFlagUtils.GetFlagString(privateAssetsFlag).Replace(',', ';'));
+            }
+
+            NuGetUIThreadHelper.JoinableTaskFactory.Run(async delegate
```

Comment:
can you explain rationale for making some of these operations async, compared with VsManagedLanguagesProjectSystemServices.cs?

Reply:
https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectSystems/CpsProjectSystem.cs#L33

Reply:
Sorry, this is a mistake.